app/gui2/src/util/ast/abstract.ts

```diff
@@ -0,0 +1,165 @@
+/**
+ * # AST
+ *
+ * The goal of the new AST is to make expression edits simple.
+ *
+ * ## Nodes
+ *
+ * A node has an ExprId and other metadata, type-specific information, and a sequence of children. Children can be node
+ * references or strings. String children are the tokens of the frontend AST. For example, an identifier node has a
+ * single string child; a parser "macro" contains strings for its reserved words and and expression references for its
+ * general expressions.
+ *
+ * Only nodes have ExprId assignments and thus are able to receive type info from the backend. Some things that are
+ * only tokens in `enso-parser` have nodes in the expression tree, e.g. operator symbols.
+ *
+ * ### Referential transparency
+ *
+ * Nodes are not dependent on any auxiliary data structure. This allows them to be easily created and rearranged.
+ *
+ * ## The AST is abstract
+ *
+ * `Tree` is concrete; this is important for preserving information such as exact whitespace characters, but it is not
+ * conducive to editing. The GUI's AST is abstract. Avoiding redundant information allows application logic to perform
+ * edits without needing to maintain unnecessary invariants. When ASTs are printed to text, any tokens needed to
+ * represent the structure are materialized (e.g. parentheses are inserted as needed.) During this printing process,
+ * the `Tree`s corresponding to preexisting AST nodes are consulted, in order to ensure that the concrete syntax of
+ * unmodified parts of the output matches the input. This is the only use of `Tree` in the frontend outside of the
+ * process of translating from `Tree` to frontend AST.
```

Comment:
I'm not sure if this will necessarily be the only use. The concrete AST might still be more suited for things like alias analysis, unless you think that it can be easily implemented on top of this abstract AST as well.